modules/ingest-common/src/main/java/org/opensearch/ingest/common/FingerprintProcessor.java

```diff
@@ -0,0 +1,261 @@
+        @Nullable List<String> fields,
+        boolean includeAllFields,
+        String targetField,
+        String hashMethod,
+        boolean ignoreMissing
+    ) {
+        super(tag, description);
+        if (fields != null) {
+            if (fields.isEmpty()) {
+                throw new IllegalArgumentException("fields cannot be empty");
+            }
+            if (fields.stream().anyMatch(Objects::isNull)) {
+                throw new IllegalArgumentException("field path cannot be null nor empty");
+            }
+            if (includeAllFields) {
+                throw new IllegalArgumentException("either fields or include_all_fields can be set");
+            }
+        } else if (!includeAllFields) {
+            throw new IllegalArgumentException("either fields or include_all_fields must be set");
+        }
+
+        if (!HASH_METHODS.contains(hashMethod.toUpperCase(Locale.ROOT))) {
+            throw new IllegalArgumentException("hash method must be MD5, SHA-1 or SHA-256");
+        }
+        this.fields = fields;
+        this.includeAllFields = includeAllFields;
+        this.targetField = targetField;
+        this.hashMethod = hashMethod;
+        this.ignoreMissing = ignoreMissing;
+    }
+
+    public List<String> getFields() {
+        return fields;
+    }
+
+    public boolean getIncludeAllFields() {
+        return includeAllFields;
+    }
+
+    public String getTargetField() {
+        return targetField;
+    }
+
+    public String getHashMethod() {
+        return hashMethod;
+    }
+
+    public boolean isIgnoreMissing() {
+        return ignoreMissing;
+    }
+
+    @Override
+    public IngestDocument execute(IngestDocument document) {
+        // we should deduplicate and sort the field names to make sure we can get consistent hash value
+        final List<String> sortedFields;
+        if (includeAllFields) {
+            Set<String> existingFields = new HashSet<>(document.getSourceAndMetadata().keySet());
+            Set<String> metadataFields = document.getMetadata()
+                .keySet()
+                .stream()
+                .map(IngestDocument.Metadata::getFieldName)
+                .collect(Collectors.toSet());
+            sortedFields = existingFields.stream().filter(field -> !metadataFields.contains(field)).sorted().collect(Collectors.toList());
+        } else {
+            sortedFields = fields.stream().distinct().sorted().collect(Collectors.toList());
+        }
+        assert (!sortedFields.isEmpty());
+
+        final StringBuilder concatenatedFields = new StringBuilder();
+        sortedFields.forEach(field -> {
+            if (!document.hasField(field)) {
+                if (ignoreMissing) {
+                    return;
+                } else {
+                    throw new IllegalArgumentException("field [" + field + "] doesn't exist");
+                }
+            }
+
+            final Object value = document.getFieldValue(field, Object.class);
```

Comment:
We may have to handle the `map` type field specially because we need to flatten and sort the sub-field names to make sure we can get consistent hash value.

Reply:
Yeah, the canonical representation does make sense, the one of the use cases I have in mind is `flat_object`: this is essentially a JSON which ideally should also be represented in canonical form (to have consistent hash output), how would we deal with that?

Reply:
During the ingesting process, actually we don't know the type of the field, the value of the `flat_object` field is just a map, so we treat these fields the same as the `object` field.